File: twa_framework.py
"""
Truncated Wigner Approximation (TWA) Framework for Dissipative Spin Systems

Implements the user-friendly TWA method from:
"User-Friendly Truncated Wigner Approximation for Dissipative Spin Dynamics"
Hosseinabadi, Chelpanova, and Marino, PRX Quantum 6, 030344 (2025)

This module provides efficient simulation of dissipative quantum many-body dynamics
using semiclassical methods with noise.
"""

import numpy as np
from typing import Callable, Dict, List, Tuple
from dataclasses import dataclass


@dataclass
class DissipationChannel:
    """Represents a dissipation channel with jump operators and rates."""
    type: str  # 'decay', 'pumping', 'dephasing'
    rate: float  # γ for decay/pumping, κ for dephasing
    qubits: List[int]  # Which qubits this channel affects


class TWASpinSimulator:
    """
    Truncated Wigner Approximation simulator for dissipative spin systems.

    Implements the protocol from Section III.A of Hosseinabadi et al. (2025):
    1. Replace spin operators with classical variables s^α_k
    2. Derive equations of motion from effective Hamiltonian H̃ = H - i∑(L̄ᵢΦᵢ - Φ̄ᵢLᵢ)
    3. Add noise terms: Φᵢ = (1/2)∑ⱼ Γᵢⱼ Lⱼ + (1/2)ξᵢ
    4. Evolve classical trajectories with stochastic noise
    5. Average over trajectories to get quantum expectation values

    Key features:
    - Conserves spin length |s_k|² for each trajectory
    - Supports spin loss, pumping, and dephasing
    - Discrete TWA (DTWA) sampling for initial conditions
    - Hardware-realistic dissipation rates (T1, T2)
    """

    def __init__(self, n_qubits: int, n_trajectories: int = 1000):
        """
        Initialize TWA simulator.

        Args:
            n_qubits: Number of qubits (spin-1/2 particles)
            n_trajectories: Number of stochastic trajectories to average
        """
        self.n_qubits = n_qubits
        self.n_trajectories = n_trajectories
        self.dissipation_channels = []

        # Pauli matrices (for reference, not used in classical evolution)
        self.pauli_x = np.array([[0, 1], [1, 0]], dtype=complex)
        self.pauli_y = np.array([[0, -1j], [1j, 0]], dtype=complex)
        self.pauli_z = np.array([[1, 0], [0, -1]], dtype=complex)

    def add_dissipation(self, channel_type: str, rate: float, qubits: List[int]):
        """
        Add dissipation channel to the system.

        Args:
            channel_type: Type of dissipation ('decay', 'pumping', 'dephasing')
            rate: Dissipation rate (γ↓, γ↑, or κ)
            qubits: List of qubit indices affected by this channel
        """
        channel = DissipationChannel(channel_type, rate, qubits)
        self.dissipation_channels.append(channel)
        print(f"Added {channel_type} channel: rate={rate}, qubits={qubits}")

    def discrete_sample_initial_state(self, initial_state: str = 'ground') -> np.ndarray:
        """
        Sample initial spin configuration using Discrete TWA (DTWA).

        For spin-1/2, uses discrete distribution from Wootters (1987):
        - For |↓⟩: sample uniformly from {(±1, ±1, -1)}
        - For |↑⟩: sample uniformly from {(±1, ±1, +1)}

        Args:
            initial_state: Initial quantum state ('ground', 'excited', 'superposition')

        Returns:
            spins: Array of shape (n_qubits, 3) with classical spin vectors
        """
        spins = np.zeros((self.n_qubits, 3))

        for k in range(self.n_qubits):
            sx = np.random.choice([-1, 1])
            sy = np.random.choice([-1, 1])

            if initial_state == 'ground':
                # |↓⟩ state: sz = -1
                sz = -1
            elif initial_state == 'excited':
                # |↑⟩ state: sz = +1
                sz = 1
            elif initial_state == 'superposition':
                # |+⟩ = (|↑⟩ + |↓⟩)/√2: sample sz uniformly
                sz = np.random.choice([-1, 1])
            else:
                raise ValueError(f"Unknown initial_state: {initial_state}")

            spins[k] = np.array([sx, sy, sz])

        return spins

    def poisson_bracket_spin(self, s_alpha: np.ndarray, H_grad: np.ndarray) -> np.ndarray:
        """
        Compute Poisson bracket {s^α, H}_p = 2 ε^{αβγ} (∂H/∂s^β) s^γ

        For spins, the Poisson bracket is:
        {s^α_k, O}_p = 2 ∑_{β,γ} ε^{αβγ} (∂O/∂s^β_k) s^γ_k

        This gives the classical equations of motion analogous to quantum commutators.

        Args:
            s_alpha: Current spin component being evolved (shape: (n_qubits,))
            H_grad: Gradient ∂H̃/∂s (shape: (n_qubits, 3))

        Returns:
            Time derivative ds^α/dt
        """
        # Levi-Civita tensor ε^{αβγ}
        epsilon = np.array([
            [[0, 0, 0], [0, 0, 1], [0, -1, 0]],  # ε^{0βγ} (α=x)
            [[0, 0, -1], [0, 0, 0], [1, 0, 0]],  # ε^{1βγ} (α=y)
            [[0, 1, 0], [-1, 0, 0], [0, 0, 0]]   # ε^{2βγ} (α=z)
        ])

        # This will be implemented by the specific Hamiltonian
        raise NotImplementedError("Subclasses must implement poisson_bracket_spin")

    def equations_of_motion_decay(self, t: float, s: np.ndarray,
                                   H_gradient_func: Callable,
                                   noise: Dict[str, np.ndarray]) -> np.ndarray:
        """
        Equations of motion for spins with decay (spin loss).

        From Table I of the paper:
        ds^x/dt = {s^x, H}_p + (γ↓/2) s^x s^z + ξ^x_↓ s^z
        ds^y/dt = {s^y, H}_p + (γ↓/2) s^y s^z + ξ^y_↓ s^z
        ds^z/dt = {s^z, H}_p - (γ↓/2)(s^x s^x + s^y s^y) - (ξ^x_↓ s^x + ξ^y_↓ s^y)

        Args:
            t: Current time
            s: Spin configuration, shape (n_qubits, 3)
            H_gradient_func: Function that computes gradient of Hamiltonian
            noise: Dictionary of noise realizations

        Returns:
            ds/dt: Time derivatives, shape (n_qubits, 3)
        """
        dsdt = np.zeros_like(s)

        # Get Hamiltonian contribution (coherent dynamics)
        H_grad = H_gradient_func(s)

        # Coherent part: cross product s × (∂H/∂s)
        # {s^α, H}_p = 2 ε^{αβγ} (∂H/∂s^β) s^γ = 2 (s × ∇H)^α
        coherent_term = 2.0 * np.cross(s, H_grad)

        # Dissipative part (for each channel)
        for channel in self.dissipation_channels:
            if channel.type == 'decay':
                gamma = channel.rate
                for k in channel.qubits:
                    # Get noise for this qubit
                    xi_x = noise['decay_x'][k]
                    xi_y = noise['decay_y'][k]

                    # Dissipative terms from Table I
                    dsdt[k, 0] += (gamma / 2) * s[k, 0] * s[k, 2] + xi_x * s[k, 2]
                    dsdt[k, 1] += (gamma / 2) * s[k, 1] * s[k, 2] + xi_y * s[k, 2]
                    dsdt[k, 2] += -(gamma / 2) * (s[k, 0]**2 + s[k, 1]**2) - (xi_x * s[k, 0] + xi_y * s[k, 1])

            elif channel.type == 'dephasing':
                kappa = channel.rate
                for k in channel.qubits:
                    # Get noise for this qubit
                    eta = noise['dephasing'][k]

                    # Dephasing terms from Table I
                    dsdt[k, 0] += 2 * eta * s[k, 1]
                    dsdt[k, 1] += -2 * eta * s[k, 0]
                    # ds^z/dt = 0 for dephasing

        # Add coherent contribution
        dsdt += coherent_term

        return dsdt

    def generate_noise(self, dt: float) -> Dict[str, np.ndarray]:
        """
        Generate Gaussian noise realizations for one time step.

        For TWA, noise satisfies:
        ⟨ξ^α_i(t)⟩ = 0
        ⟨ξ^α_i(t) ξ^β_j(t')⟩ = γ_i δ_{ij} δ_{αβ} δ(t - t')

        In discrete time: ξ ~ N(0, γ/dt)

        Args:
            dt: Time step size

        Returns:
            Dictionary of noise arrays
        """
        noise = {}

        for channel in self.dissipation_channels:
            rate = channel.rate

            # Safety check: ensure rate*dt is not too large (causes instability)
            if rate * dt > 1.0:
                print(f"WARNING: rate*dt = {rate*dt:.2e} > 1, noise may be too large!")
                print(f"  Consider reducing dt or rescaling dissipation rates")

            sigma = np.sqrt(max(rate / dt, 1e-20))  # Avoid sqrt of negative/zero

            if channel.type == 'decay':
                # Two noise components (x and y) for each qubit
                noise['decay_x'] = np.random.normal(0, sigma, self.n_qubits)
                noise['decay_y'] = np.random.normal(0, sigma, self.n_qubits)

            elif channel.type == 'dephasing':
                # One noise component (z) for each qubit
                noise['dephasing'] = np.random.normal(0, sigma, self.n_qubits)

        return noise

    def rk4_step(self, t: float, s: np.ndarray, dt: float,
                 H_gradient_func: Callable, noise: Dict[str, np.ndarray]) -> np.ndarray:
        """
        Runge-Kutta 4th order integration step with Stratonovich interpretation.

        Args:
            t: Current time
            s: Current spin configuration
            dt: Time step
            H_gradient_func: Hamiltonian gradient function
            noise: Noise realization for this step

        Returns:
            s_new: Updated spin configuration
        """
        # RK4 for stochastic differential equations (Stratonovich)
        k1 = self.equations_of_motion_decay(t, s, H_gradient_func, noise)
        k2 = self.equations_of_motion_decay(t + dt/2, s + dt*k1/2, H_gradient_func, noise)
        k3 = self.equations_of_motion_decay(t + dt/2, s + dt*k2/2, H_gradient_func, noise)
        k4 = self.equations_of_motion_decay(t + dt, s + dt*k3, H_gradient_func, noise)

        s_new = s + (dt / 6) * (k1 + 2*k2 + 2*k3 + k4)

        return s_new

    def check_spin_conservation(self, s: np.ndarray, renormalize: bool = False) -> np.ndarray:
        """
        Check that spin length is conserved for each qubit.

        For TWA, |s_k|² must remain constant (= 3 for spin-1/2 with our convention).
        This is automatically satisfied if equations come from effective Hamiltonian.

        Args:
            s: Spin configuration
            renormalize: If True, renormalize spins to correct length (safety mechanism)

        Returns:
            Array of spin lengths squared (before renormalization if applied)
        """
        s_squared = np.sum(s**2, axis=1)

        if renormalize:
            # Target spin length squared (3 for spin-1/2)
            target = 3.0
            for k in range(len(s)):
                current_norm = s_squared[k]
                if current_norm > 1e-10:  # Avoid division by zero
                    if current_norm > 10.0 or current_norm < 0.1:
                        # Spin norm has deviated significantly - renormalize
                        s[k] *= np.sqrt(target / current_norm)

        return s_squared

    def compute_expectation_values(self, spins_all_trajectories: np.ndarray) -> Dict[str, np.ndarray]:
        """
        Compute expectation values by averaging over trajectories.

        ⟨σ̂^α_k(t)⟩ ≈ (1/N_tr) ∑_n s^α_{k,n}(t)

        Args:
            spins_all_trajectories: Array of shape (n_trajectories, n_qubits, 3)

        Returns:
            Dictionary with expectation values
        """
        expectations = {
            'sx': np.mean(spins_all_trajectories[:, :, 0], axis=0),
            'sy': np.mean(spins_all_trajectories[:, :, 1], axis=0),
            'sz': np.mean(spins_all_trajectories[:, :, 2], axis=0),
            'magnetization': np.mean(np.sum(spins_all_trajectories[:, :, 2], axis=1)),
        }
        return expectations


def apply_discrete_sampling_correction(expectations: Dict[str, np.ndarray]) -> Dict[str, np.ndarray]:
    """
    Apply corrections for discrete sampling artifacts.

    Discrete TWA can introduce systematic errors compared to continuous sampling.
    For single-point expectation values, the error is typically small (< 1%).

    Args:
        expectations: Raw expectation values from TWA

    Returns:
        Corrected expectation values
    """
    # For spin-1/2 with DTWA, no correction needed for ⟨σ^z⟩
    # Corrections would be needed for higher-order correlations
    return expectations


# Hardware-realistic dissipation rates for 171Yb+
class IonTrapDissipationRates:
    """
    Realistic dissipation rates for 171Yb+ trapped ions.

    IMPORTANT: Rates are rescaled to match Hamiltonian energy units.
    For quantum chemistry simulations in a.u., use energy_scale parameter.
    """

    def __init__(self, energy_scale: float = 1.0):
        """
        Initialize dissipation rates.

        Args:
            energy_scale: Typical energy scale of Hamiltonian (for unit matching)
                         For H2O model: ~10 Hartree
                         For real atomic units: use 1.0
        """
        # Hardware specifications for 171Yb+ trapped ions (SI units)
        self.T1_SI = 1000.0  # seconds (effectively infinite)
        self.T2_SI = 1.0     # seconds (dephasing time)

        # Convert to rates in Hz
        self.gamma_decay_SI = 1.0 / self.T1_SI  # Hz (spin loss rate)
        self.kappa_dephasing_SI = 1.0 / self.T2_SI  # Hz (dephasing rate)

        # Atomic unit conversions
        # 1 a.u. time = ℏ/E_h = 2.4189×10^-17 s
        # 1 a.u. energy = E_h = 27.211 eV = 4.3597×10^-18 J
        self.AU_TIME = 2.4189e-17  # seconds per atomic time unit
        self.AU_ENERGY = 27.211    # eV

        # Convert to atomic units (extremely small for long coherence times)
        self.gamma_decay_au = self.gamma_decay_SI * self.AU_TIME
        self.kappa_dephasing_au = self.kappa_dephasing_SI * self.AU_TIME

        # Rescale to match Hamiltonian energy units
        # If H ~ 10 Hartree, time scale is ~0.1 a.u., so rates should scale accordingly
        self.energy_scale = energy_scale
        self.gamma_decay = self.gamma_decay_au * energy_scale
        self.kappa_dephasing = self.kappa_dephasing_au * energy_scale

        # Gate error rates
        self.single_qubit_error = 1 - 0.998  # 0.2% error
        self.two_qubit_error = 1 - 0.970     # 3.0% error

        print(f"\nDissipation rates initialized:")
        print(f"  T1 = {self.T1_SI} s → γ = {self.gamma_decay:.2e} (scaled)")
        print(f"  T2 = {self.T2_SI} s → κ = {self.kappa_dephasing:.2e} (scaled)")
        print(f"  Energy scale: {energy_scale:.2f}")
        print(f"  (Rates scaled by {energy_scale} to match Hamiltonian units)")

    def effective_decay_during_gate(self, gate_time: float, gate_type: str = 'two_qubit') -> float:
        """
        Compute effective decay during gate operation.

        Args:
            gate_time: Gate duration in seconds
            gate_type: 'single_qubit' or 'two_qubit'

        Returns:
            Effective decay probability
        """
        if gate_type == 'single_qubit':
            # ~1-10 μs gate time
            return gate_time / self.T1 + self.single_qubit_error
        else:
            # ~100 μs - 1 ms gate time
            return gate_time / self.T1 + self.two_qubit_error


if __name__ == "__main__":
    print("=" * 70)
    print("TWA FRAMEWORK MODULE")
    print("=" * 70)
    print()
    print("This module provides the Truncated Wigner Approximation framework")
    print("for simulating dissipative spin dynamics in trapped-ion systems.")
    print()
    print("Key features:")
    print("  - Dissipative dynamics with T1/T2 decoherence")
    print("  - Discrete TWA (DTWA) sampling")
    print("  - Hardware-realistic 171Yb+ parameters")
    print("  - Conserves spin length for each trajectory")
    print()
    print("Import this module and use TWASpinSimulator class for simulations.")
    print("=" * 70)

File: cpp/twa_framework.cpp
/**
 * Truncated Wigner Approximation (TWA) Framework - C++ Implementation
 */

#include "twa_framework.hpp"
#include <iostream>
#include <cmath>
#include <algorithm>
#include <numeric>

namespace twa {

// Spin3D methods
void Spin3D::normalize(double target) {
    double norm_sq = norm_squared();
    if (norm_sq > 1e-10) {
        double scale = std::sqrt(target / norm_sq);
        x *= scale;
        y *= scale;
        z *= scale;
    }
}

// TWASpinSimulator implementation
TWASpinSimulator::TWASpinSimulator(int n_qubits, int n_trajectories)
    : n_qubits_(n_qubits),
      n_trajectories_(n_trajectories),
      rng_(std::random_device{}()),
      normal_dist_(0.0, 1.0),
      sign_dist_(0, 1) {

    std::cout << "TWA Spin Simulator initialized:" << std::endl;
    std::cout << "  Qubits: " << n_qubits_ << std::endl;
    std::cout << "  Trajectories: " << n_trajectories_ << std::endl;
}

void TWASpinSimulator::add_dissipation(ChannelType type, double rate,
                                       const std::vector<int>& qubits) {
    dissipation_channels_.emplace_back(type, rate, qubits);

    std::string type_str;
    switch(type) {
        case ChannelType::DECAY: type_str = "decay"; break;
        case ChannelType::PUMPING: type_str = "pumping"; break;
        case ChannelType::DEPHASING: type_str = "dephasing"; break;
    }

    std::cout << "Added " << type_str << " channel: rate=" << rate
              << ", qubits=[";
    for (size_t i = 0; i < qubits.size(); ++i) {
        std::cout << qubits[i];
        if (i < qubits.size() - 1) std::cout << ", ";
    }
    std::cout << "]" << std::endl;
}

void TWASpinSimulator::clear_dissipation() {
    dissipation_channels_.clear();
}

std::vector<Spin3D> TWASpinSimulator::discrete_sample_initial_state(
    const std::string& state) {

    std::vector<Spin3D> spins(n_qubits_);

    for (int k = 0; k < n_qubits_; ++k) {
        double sx = (sign_dist_(rng_) == 0) ? -1.0 : 1.0;
        double sy = (sign_dist_(rng_) == 0) ? -1.0 : 1.0;
        double sz;

        if (state == "ground") {
            sz = -1.0;
        } else if (state == "excited") {
            sz = 1.0;
        } else if (state == "superposition") {
            sz = (sign_dist_(rng_) == 0) ? -1.0 : 1.0;
        } else {
            std::cerr << "Unknown initial state: " << state << std::endl;
            sz = -1.0;
        }

        spins[k] = Spin3D(sx, sy, sz);
    }

    return spins;
}

NoiseData TWASpinSimulator::generate_noise(double dt) {
    NoiseData noise(n_qubits_);

    for (const auto& channel : dissipation_channels_) {
        double rate = channel.rate;

        // Safety check
        if (rate * dt > 1.0) {
            std::cout << "WARNING: rate*dt = " << rate*dt
                     << " > 1, noise may be too large!" << std::endl;
        }

        double sigma = std::sqrt(std::max(rate / dt, 1e-20));

        if (channel.type == ChannelType::DECAY) {
            for (int k = 0; k < n_qubits_; ++k) {
                noise.decay_x[k] = normal_dist_(rng_) * sigma;
                noise.decay_y[k] = normal_dist_(rng_) * sigma;
            }
        } else if (channel.type == ChannelType::DEPHASING) {
            for (int k = 0; k < n_qubits_; ++k) {
                noise.dephasing[k] = normal_dist_(rng_) * sigma;
            }
        }
    }

    return noise;
}

std::vector<Spin3D> TWASpinSimulator::equations_of_motion(
    [[maybe_unused]] double t,
    const std::vector<Spin3D>& spins,
    const HamiltonianGradientFunc& grad_func,
    const NoiseData& noise) {

    std::vector<Spin3D> dsdt(n_qubits_);

    // Coherent part: {s, H}_p = 2(s × ∇H)
    auto grad_H = grad_func(spins);

    for (int k = 0; k < n_qubits_; ++k) {
        dsdt[k] = spins[k].cross(grad_H[k]) * 2.0;
    }

    // Dissipative part
    for (const auto& channel : dissipation_channels_) {
        if (channel.type == ChannelType::DECAY) {
            double gamma = channel.rate;

            for (int k : channel.qubits) {
                if (k < n_qubits_) {
                    double xi_x = noise.decay_x[k];
                    double xi_y = noise.decay_y[k];
                    double sz = spins[k].z;
                    double sx = spins[k].x;
                    double sy = spins[k].y;

                    // From TWA paper Table I (spin loss channel)
                    dsdt[k].x += (gamma / 2.0) * sx * sz + xi_x * sz;
                    dsdt[k].y += (gamma / 2.0) * sy * sz + xi_y * sz;
                    dsdt[k].z += -(gamma / 2.0) * (sx*sx + sy*sy) - (xi_x * sx + xi_y * sy);
                }
            }
        } else if (channel.type == ChannelType::DEPHASING) {
            for (int k : channel.qubits) {
                if (k < n_qubits_) {
                    double eta = noise.dephasing[k];

                    // From TWA paper Table I (dephasing channel)
                    dsdt[k].x += 2.0 * eta * spins[k].y;
                    dsdt[k].y += -2.0 * eta * spins[k].x;
                    // dsdt[k].z unchanged by pure dephasing
                }
            }
        }
    }

    return dsdt;
}

std::vector<Spin3D> TWASpinSimulator::rk4_step(
    double t,
    const std::vector<Spin3D>& spins,
    double dt,
    const HamiltonianGradientFunc& grad_func,
    const NoiseData& noise) {

    // RK4 for stochastic differential equations (Stratonovich)
    auto k1 = equations_of_motion(t, spins, grad_func, noise);

    std::vector<Spin3D> s_temp(n_qubits_);
    for (int k = 0; k < n_qubits_; ++k) {
        s_temp[k] = spins[k] + k1[k] * (dt / 2.0);
    }
    auto k2 = equations_of_motion(t + dt/2.0, s_temp, grad_func, noise);

    for (int k = 0; k < n_qubits_; ++k) {
        s_temp[k] = spins[k] + k2[k] * (dt / 2.0);
    }
    auto k3 = equations_of_motion(t + dt/2.0, s_temp, grad_func, noise);

    for (int k = 0; k < n_qubits_; ++k) {
        s_temp[k] = spins[k] + k3[k] * dt;
    }
    auto k4 = equations_of_motion(t + dt, s_temp, grad_func, noise);

    std::vector<Spin3D> s_new(n_qubits_);
    for (int k = 0; k < n_qubits_; ++k) {
        s_new[k] = spins[k] + (k1[k] + k2[k] * 2.0 + k3[k] * 2.0 + k4[k]) * (dt / 6.0);
    }

    return s_new;
}

void TWASpinSimulator::check_spin_conservation(std::vector<Spin3D>& spins,
                                              bool renormalize) {
    if (!renormalize) return;

    constexpr double target = 3.0;  // |s|² for spin-1/2

    for (auto& spin : spins) {
        double norm_sq = spin.norm_squared();
        if (norm_sq > 10.0 || norm_sq < 0.1) {
            spin.normalize(target);
        }
    }
}

// IonTrapDissipationRates implementation
IonTrapDissipationRates::IonTrapDissipationRates(double energy_scale_)
    : T1_SI(1000.0),
      T2_SI(1.0),
      energy_scale(energy_scale_),
      single_qubit_error(1.0 - 0.998),
      two_qubit_error(1.0 - 0.970) {

    // Convert to rates in Hz
    gamma_decay_SI = 1.0 / T1_SI;
    kappa_dephasing_SI = 1.0 / T2_SI;

    // Convert to atomic units
    gamma_decay_au = gamma_decay_SI * AU_TIME;
    kappa_dephasing_au = kappa_dephasing_SI * AU_TIME;

    // Rescale to match Hamiltonian energy units
    gamma_decay = gamma_decay_au * energy_scale;
    kappa_dephasing = kappa_dephasing_au * energy_scale;
}

void IonTrapDissipationRates::print_info() const {
    std::cout << "\nDissipation rates initialized:" << std::endl;
    std::cout << "  T1 = " << T1_SI << " s → γ = " << gamma_decay << " (scaled)" << std::endl;
    std::cout << "  T2 = " << T2_SI << " s → κ = " << kappa_dephasing << " (scaled)" << std::endl;
    std::cout << "  Energy scale: " << energy_scale << std::endl;
}

} // namespace twa
